lib/Sema/ConstraintSystem.cpp

```diff
@@ -1175,7 +1175,8 @@ ConstraintSystem::getTypeOfMemberReference(
          func->hasArchetypeSelf())) {
       openedType = openedType->replaceCovariantResultType(
                      baseObjTy,
-                     func->getNumParameterLists());
+                     func->getNumParameterLists(),
+                     !baseObjTy->getAnyOptionalObjectType());
```

Comment:
It's perfectly fine to end up with `Optional<ImplicitlyUnwrappedOptional<T>>` if we have a situation like this:

    protocol P {
    }

   extension P {
     func f() -> Self? { return self }
   }

  extension ImplicitlyUnwrappedOptional: P { }

  var iuo: Int! = 1
  iuo.f() // this is okay! returns an Optional<ImplicitlyUnwrappedOptional<Int>>

So the trick is to unwrap the IUO to get to the right base type.